yalafi/packages/xspace.py

```diff
@@ -0,0 +1,37 @@
+
+#
+#   YaLafi module for LaTeX package xspace
+#
+
+from yalafi import defs
+from yalafi.defs import Macro, InitModule
+
+require_packages = []
+
+def init_module(parser, options):
+    parms = parser.parms
+
+    macros_latex = ""
+
+    macros_python = [
+
+        Macro(parms, '\\xspace', args='OA', repl=h_xspace),

```

Comment:
The LaTeX command `xspace` does not really have parameters, but here the code is claiming two (`OA`). Is this ok? It seems to work in my tests, but I am unsure if this “consumes” too much of what comes after `\xspace`.

Reply:
Thanks a lot! Your method of inserting a new token seems to be a clever solution. I need to think about it twice.

Reply:
If there is another more straightforward way to insert a new token, please let me know :) This hack with `OA` was what I came up with after some print debugging, and inspiration by the handling of `phantom`.

Reply:
Your scheme should work well in principle. There are, however, some minor problems. For the beginning, I'd propose
```
        Macro(parms, '\\xspace', args='A', repl=h_xspace),
```
and
```
def h_xspace(parser, buf, mac, args, pos):
    arg = args[0]
    if len(arg) == 1 and type(arg[0]) is defs.VoidToken:
        return []
    if not arg[0].txt in xspace_excl:
        arg.insert(0, defs.SpecialToken(arg[0].pos, '\\;'))
    return arg
```
This does work with your tests.
- It seems that the optional argument can be omitted.
- If you request a mandatory argument, then the corresponding entry in `args` will contain at least one token. If the argument is empty (e.g., on \{\} or a paragraph break), `defs.VoidToken` is passed.
- Every token has an attribute `txt`.
- By the way, the line defining `xspace_excl` currently is longer than 80 characters.

Unfortunately, this and your current version reveal a bug in `Parser.expand_arguments()` in yalafi/parser.py. On `A{\xspace}B`, we get `A B`, since the closing \} is mistakenly assumed as argument of \\xspace. Instead, we should receive a VoidToken. I will raise a 'bug' issue.

Additionally, I was thinking about a case like
```
 \xspace {xyz\newcommand\y{Y}}
```
Firstly, the outer \{\} braces are removed during argument detection for h\_xspace(). This pushes back
```
\;xyz\newcommand\y{Y}
```
such that \\y is not restricted to the original scope anymore. I think, we can accept that. YaLafi doesn't account for scoping, anyway.
Secondly, the real \\xspace does *not* introduce space, if the next token is a '{'.
In summary, we need an information, if braces were present around the mandatory argument(s) of functions like h\_xspace(). I will raise an 'enhancement' issue for that, too.

What do you think? I would propose to first fix the bug with seeing '}' as argument, and give you a ring. After updating yalafi/parser.py, you could check that everything works. We could merge the PR and leave the problem due to a \{\}-braced argument "for further work". (The fix of the second issue will probably add an argument to the handler function interface and affect a lot of Python files. But I think, it is worth to enhance the interface with the additional information.)